treeherder/etl/taskcluster_pulse/handler.py

```diff
@@ -0,0 +1,340 @@
+    # This will only work if the previous run has not yet been processed by Treeherder
+    # since _remove_existing_jobs() will prevent it
+    if message["payload"]["runId"] > 0:
+        jobs.append(await handleTaskRerun(parsedRoute, task, message["payload"]))
+
+    if not taskType:
+        raise Exception("Unknown exchange: {exchange}".format(exchange=message["exchange"]))
+    elif taskType == "pending":
+        jobs.append(handleTaskPending(parsedRoute, task, message["payload"]))
+    elif taskType == "running":
+        jobs.append(handleTaskRunning(parsedRoute, task, message["payload"]))
+    elif taskType in ("completed", "failed"):
+        jobs.append(await handleTaskCompleted(parsedRoute, task, message["payload"]))
+    elif taskType == "exception":
+        jobs.append(await handleTaskException(parsedRoute, task, message["payload"]))
+
+    return jobs
+
+
+# Builds the basic Treeherder job message that's universal for all
+# messsage types.
+#
+# Specific handlers for each message type will add/remove information necessary
+# for the type of task event..
+def buildMessage(pushInfo, task, runId, message):
+    taskId = message["status"]["taskId"]
+    jobRun = message["status"]["runs"][runId]
+    treeherderConfig = task["extra"]["treeherder"]
+
+    job = {
+      "buildSystem": "taskcluster",
+      "owner": task["metadata"]["owner"],
+      "taskId": "{taskId}/{runId}".format(taskId=slugid.decode(taskId), runId=runId),
+      # TODO: In a follow up we will use the right name - bug 1560596
+      "realTaskId": taskId,
+      "runId": runId,
+      "isRetried": False,
+      "display": {
+        # jobSymbols could be an integer (i.e. Chunk ID) but need to be strings
+        # for treeherder
+        "jobSymbol": str(treeherderConfig["symbol"]),
+        "groupSymbol": treeherderConfig.get("groupSymbol", "?"),
+        # Maximum job name length is 100 chars...
+        "jobName": task["metadata"]["name"][0:99],
+      },
+      "state": stateFromRun(jobRun),
+      "result": resultFromRun(jobRun),
+      "tier": treeherderConfig.get("tier", 1),
+      "timeScheduled": task["created"],
+      "jobKind": treeherderConfig.get("jobKind", "other"),
+      "reason": treeherderConfig.get("reason", "scheduled"),
+      "jobInfo": {
+        "links": [],
+        "summary": task["metadata"]["description"],
+      },
+    }
+
+    job["origin"] = {
+      "kind": pushInfo["origin"],
+      "project": pushInfo["project"],
+      "revision": pushInfo["revision"],
+    }
+
+    if pushInfo["origin"] == "hg.mozilla.org":
+        job["origin"]["pushLogID"] = pushInfo["id"]
+    else:
+        job["origin"]["pullRequestID"] = pushInfo["id"]
+        job["origin"]["owner"] = pushInfo["owner"]
+
+    # Transform "collection" into an array of labels if task doesn't
+    # define "labels".
+    labels = treeherderConfig.get("labels", [])
+    if not labels:
+        if not treeherderConfig.get("collection"):
+            labels = ["opt"]
```

Comment:
This is probably fine, and is consistent with the original code for taskcluster-treeherder, iirc.  But I seem to recall having a conversation with @jmaher at some point where he thought the default fall-back should be something other than opt.  I could be mis-remembering, but might be worth checking in with him.  :)

Reply:
there are a lot of different options, mostly I am concerned about the array of options like ccov and asan which might not be worrisome here.  Could we log when we set this by default so we could handle this edge case more gracefully?

Reply:
Hi @jmaher 
At this point any more changes for this I would like to put it to [bug 1560596](https://bugzilla.mozilla.org/show_bug.cgi?id=1560596)